RIOT-OS/RIOT
# sys/entropy_source: add new module for entropy sources

### Contribution description

This PR adds an `entropy_source` module which acts as a place to implement different entropy sources as submodules. Currently, an ADC based source and a zero entropy source (for testing only!) are implemented. Additionally, a common component provides functionality that is usable for all. Sources can optionally enable health tests (executed during runtime) and output conditioning.

Output values can be used directly to create low-requirement seeds, or they can be used to feed more advanced "high level" entropy handlers that accumulate values of multiple entropy sources. More information can be found in the [module documentation](https://github.com/PeterKietzmann/RIOT/blob/pr_add_entropy_source/sys/entropy_source/doc.txt). 

I have an other entropy source lying around that bases on jitters between differently  clocked timers, that I will PR afterwards. As an immediate next step, I will provide a PR that adds an advanced entropy module which deals with multiple sources and can be used for cryptographic purposes (seed generation).

### Open Aspects
I marked the PR as WIP because some aspects are not yet decided and implemented:
- ~~Expose configuration options to kconfig.~~
- Do we want to run startup tests on module initialization?
- ~~Do we need to return the "estimated" amount of entropy per request? The estimate would fully base on the parameters given by configuration.~~
- Do we want a more advanced conditioning component next to the *van Neumann* extractor? An alternative would be a cryptographic hash function. This might unnecessarily bloat the module  for non-crypto purposes and it might be duplicate when used with cryptographic "high level" entropy implementations that include cryptographic derivation functions anyway.

### Testing procedure

- Read and understand the [high level documentation](https://github.com/PeterKietzmann/RIOT/blob/pr_add_entropy_source/sys/entropy_source/doc.txt).
- `make doc` build the doxygen documentation.
- Execute the provided test on a BOARD and on native and compare the results with the expected results in the README.
- Play around with the default configuration values [here](https://github.com/PeterKietzmann/RIOT/blob/pr_add_entropy_source/sys/include/entropy_source/adc_noise.h#L43) and [here](https://github.com/PeterKietzmann/RIOT/blob/pr_add_entropy_source/sys/include/entropy_source/zero_entropy.h). Re-run the test. For the ADC source, connect an exposed  ADC pin to Vcc or GND and see how the error messages behave. Eventually increase the number of dumped samples, see how the output values are affected. With conditioning enabled, see how the measured runtime is affected.

### Issues/PRs references
#3425, #4294, #5153
